src/ispypsa/translator/temporal_filters.py

```diff
@@ -0,0 +1,182 @@
+from datetime import datetime, timedelta
+
+import pandas as pd
+
+from ispypsa.config.validators import ModelConfig
+
+
+def apply_snapshot_filter(time_series_data, snapshot):
+    """Filters a timeseries pandas DataFrame based using the datetime values in
+     the snapshot index.
+
+    Examples:
+
+    >>> datetime_index = pd.date_range('2020-01-01', '2020-01-03', freq='H')
+    >>> time_series_data = pd.DataFrame({'Datetime': datetime_index, 'Value': range(len(datetime_index))})
+    >>> snapshot = pd.DataFrame(index=datetime_index[::12])  # Every 12 hours
+    >>> apply_snapshot_filter(time_series_data, snapshot)
+                  Datetime  Value
+    0  2020-01-01 00:00:00      0
+    12 2020-01-01 12:00:00     12
+    24 2020-01-02 00:00:00     24
+    36 2020-01-02 12:00:00     36
+    48 2020-01-03 00:00:00     48
+
+    """
+    return time_series_data[time_series_data["Datetime"].isin(snapshot.index)]
+
+
+def filter_snapshot(config: ModelConfig, snapshot: pd.DataFrame):
+    """Appy filter to the snapshot based on the model config.
+
+    - If config.temporal.representative_weeks is not None then filter the
+      snapshot based on the supplied list of representative weeks.
+
+    Examples:
+
+    # Create dummy config class with just data need for example.
+
+    >>> from dataclasses import dataclass
+
+    >>> @dataclass
+    ... class Temporal:
+    ...     representative_weeks: list[int]
+    ...     start_year: int
+    ...     end_year: int
+    ...     year_type: str
+
+    >>> @dataclass
+    ... class ModelConfig:
+    ...     temporal: Temporal
+
+    >>> config = ModelConfig(
+    ... Temporal(
+    ...     representative_weeks=[0],
+    ...     start_year=2024,
+    ...     end_year=2024,
+    ...     year_type='calendar'
+    ... ))
+
+    >>> snapshot = pd.DataFrame(index=pd.date_range('2024-01-01', '2024-12-31', freq='H'))
+
+    >>> snapshot = filter_snapshot(config, snapshot)
+
+    >>> snapshot.index[0]
+    Timestamp('2024-01-01 01:00:00')
+
+    >>> snapshot.index[-1]
+    Timestamp('2024-01-08 00:00:00')
+
+    Args:
+         config: ModelConfig defining snapshot filtering.
+         snapshot: pd.DataFrame with datetime index containing the snapshot
+    """
+    if config.temporal.representative_weeks is not None:
+        snapshot = filter_snapshot_for_representative_weeks(
+            representative_weeks=config.temporal.representative_weeks,
+            snapshot=snapshot,
+            start_year=config.temporal.start_year,
+            end_year=config.temporal.end_year,
+            year_type=config.temporal.year_type,
+        )
+    return snapshot
+
+
+def filter_snapshot_for_representative_weeks(
+    representative_weeks: list[int],
+    snapshot: pd.DataFrame,
```

Comment:
As per other comments about naming, is the snapshot the "filtered" timeseries fed to PyPSA or the complete time series that is filtered before being fed into PyPSA? I think this arg name should be changed

Reply:
I think it OK to leave as is.